### DISCUSSION TITLE
Is there a way to skip uninstaller?

### DISCUSSION THREAD

**Discussion by Xelbayria** [ScoopInstaller/Scoop#6159] [Help and Questions]
Today, I just tried to update a software using scoop and got an error: `uninstaller.exe cannot be found` something like that. Before that, i had no problem. 

Is there a way to skip the uninstaller and just update the software because the software will uninstall the current package and install a new version (no different from updating). 

I've looked at the wiki and cannot find the answer. the JSON is using scripts to run the installer to update. There should be a script that prevent using uninstaller or something?

**Comment by HUMORCE** [comment]
which app?

**Reply by Xelbayria** [reply]
input-director

here's my JSON: 
```json
{
    "version": "2.3",
    "description": "Input Director enables the control of multiple Windows systems using the keyboard/mouse attached to one computer",
    "homepage": "https://www.inputdirector.com/",
    "license": "Freeware",
    "notes": "If uninstalled, then you must reboot to complete the uninstallation.",
    "url": "https://www.inputdirector.com/downloads/InputDirector.v2.3.zip",
    "hash": "51316c7993058fb7973f2a24523aabe7da66c08e89511486fc924f030c9af538",
    "pre_install": "Get-ChildItem \"$dir\" '*.Setup.exe' | Rename-Item -NewName 'InputDirector.Setup.exe'",
    "installer": {
        "script": [
            "$install_dir = $dir -replace '%', '^%' # escape any '%' in the path",
            "Set-Content -Encoding Ascii -Path \"$dir\\install.bat\" -Value @\"",
            "taskkill /IM \"InputDirector.exe\" /IM \"IDWinService.exe\" /IM \"InputDirectorSessionHelper.exe\" /F",
            "@echo Running Installer",
            "@`\"$install_dir\\InputDirector.Setup.exe`\" /S",
            "\"@",
            "Invoke-ExternalCommand 'cmd.exe' -Args @('/C', \"`\"$install_dir\\install.bat`\"\") -RunAs | Out-Null",
            "Remove-Item \"$dir\\install.bat\"",
            "Remove-Item \"$dir\\InputDirector.Setup.exe\""
        ]
    },
    "uninstaller": {
        "file": "uninstall.exe"
    },
    "checkver": {
        "url": "https://www.inputdirector.com/downloads.html",
        "regex": "Download Input Director v([\\d.]+)"
    },
    "autoupdate": {
        "url": "https://www.inputdirector.com/downloads/InputDirector.v$version.zip"
    }
}
```

the reason why it can't find "uninstall.exe" is beacuse it's in `C:\Program Files\Input Director\...`, I only use Scoop just to update it. it's simple. 

EDIT: 
one thing to note: I don't need scoop to uninstall the software, i just need the installer.exe to take care of that. Because if I use scoop's uninstaller, then I would have to reboot my computer to complete the uninstallation.

**Reply by HUMORCE** [reply]
```
    "uninstaller": {
        "script": [
            "if ($cmd -eq 'uninstall') { <UNINSTALLATION...> }"
        ]
    },
```

**Reply by Xelbayria** [reply]
Here's the error I am talking about "uninstaller not being found" 

`Uninstaller ~\Scoop\apps\inputdirector-np\2.2\uninstall.exe is missing.`

I've tried that code and it doesn't skip the uninstaller procedure. you'll need to explain what's the purpose of that code.

**Reply by HUMORCE** [reply]
Specify a full path of `uninstall.exe`.

If you allow the installer to install the app to a directory that is not managed by Scoop, then you should track the path of `uninstall.exe` and write to `$dir`(i.e. `scoop\apps\<app>\<version|current>`) while installing.

> you'll need to explain what's the purpose of that code.

With `if ($cmd -eq 'uninstall') { <UNINSTALLATION...> }`, <UNINSTALLATION...> will perform when `scoop uninstall <app>` only, instead `scoop update <app>` will not trigger uninstall script.

**Reply by Xelbayria** [reply]
I see. I'll run another test to see If i can get `scoop update <app>` skipping uninstall stuff and working properly.

**Reply by Xelbayria** [reply]
Is there a way to not use $dir. just directly use the path to target the `uninstaller.exe`?

**Reply by HUMORCE** [reply]
answered last reply.

**Reply by Xelbayria** [reply]
Im sorry, i don't understand what you mean via your answer. that's why I asked the question.

**Reply by HUMORCE** [reply]
You need to record path of `uninstaller.exe` at installation step, then you can reuse the path at uninstall script because the manifest is a non-portable installation, files does not extract to Scoop managed directory(e.g. `<Scoop>\apps\<app>\current`). 

It's how you use powershell, not Scoop.

for example:

```
installer:
  script:
    $target = "d:\app"
    installer.exe -target d:\app
    "$target\uninstaller.exe" > $dir\uninstaller.path

uninstaller:
   script:
     $uninstaller_path = Get-Content $dir\uninstaller.path
     Invoke-Expression "$uninstaller_path --silent"
```

**Reply by Xelbayria** [reply]
that's how it is. I understand now.

**Reply by Xelbayria** [reply]
i included this `"if ($cmd -eq 'uninstall') { <UNINSTALLATION...> }"` into the code with the script above? it should skip the uninstalling part. and just update, correct? 

I tried the script and it required Administrator to run it.   the script of "uninstaller" is still including the $dir   for example `d:\app` would become `d:\scoop\app\inputdirector-np\app\uninstaller.exe`

that shouldn't have include the $dir. and just use `d:\app` as it is.

**Reply by HUMORCE** [reply]
> i included this "if ($cmd -eq 'uninstall') { <UNINSTALLATION...> }" into the code with the script above? it should skip the uninstalling part. and just update, correct?

If it doesn't work in `uninstaller.script`, then `pre_uninstall` instead.

> I tried the script and it required Administrator to run it.

The program's uninstaller requires administrator privileges, not Scoop.

>  the script of "uninstaller" is still including the $dir for example d:\app would become d:\scoop\app\inputdirector-np\app\uninstaller.exe

In the example I provided, the actual path of `uninstaller.exe` is written to file `$dir\uninstaller.path`.

If you allow the user to manually select the actual installation path, then you need to extract it from the installer's installation log, or if the installer doesn't support log output, then you need to get it by other method.

**Reply by Xelbayria** [reply]
I know about `$dir\uninstaller.path`. it's a good way to get the location of uninstall.exe

> The program's uninstaller requires administrator privileges

i don't need it to uninstall the app, right? that's why this `"if ($cmd -eq 'uninstall') { <UNINSTALLATION...> }"` will tell scoop to skip it, right? 

I'll try `pre_uninstall`,
